services/worker-manager/providers.md

```diff
@@ -0,0 +1,117 @@
+
+The provider's `createWorker` method is called with `workerPool` (an instance of the WorkerPool Azure entity class), `workerGroup`, `workerId`, and `input`.
+The `input` matches the `create-worker-request.yml` schema, and that schema can be adjusted to allow provider-specific parameters.
+The return value should be an instance of the Worker azure entity class.
+[Idempotency](../../dev-docs/idempotency.md) of this method is the responsibilty of the provider.
+
+The provider's `removeWorker` method is called with an instance of the Worker Azure entity class.
+There are no restrictions on the state of that instance on return: it may still exist, and even have state RUNNING.
+
+#### Registering
+
+The provider's `registerWorker` method is called as a part of `registerWorker`, given both the WorkerPool and Worker instances, as well as the `workerIdentityProof` from the API request.
+The API method first verifies that the given worker exists in the Azure table and that its worker pool and provider ID match.
+
+The provider should verify the supplied identity proof and, if it is valid, modify the worker entry as appropriate (at least setting its state to RUNNING).
+Providers that wish to limit registration to once per worker should return an error message from this function if the worker is already RUNNING.
+If the method returns sucessfully, then the caller will be given Taskcluster credentials appropriate to the worker.
+
+This API method is not scope-protected, and implements an access-control mechanism based on the identity proof.
+It should be written with defensive programming in mind.
+Check all inputs carefully and reject the request if anything is incorrect.
+Do not reveal any information to a potential attacker in error messages -- avoid even indicating what portion of the input is incorrect (to prevent guessing attacks).
+
+### Provider Data
+
+Each worker entry has a `providerData` property which can be used to store arbitrary data about the worker.
+The format of this object is entirely at the discretion of the provider.
+Note that Azure entities have a fixed maximum size, so it is best to avoid storing any data in this property that grows without bound.
+
+## Provisioning
+
+The worker-manager runs a single "provisioner" process that runs the provisioning loop, where new workers are created to execute queued tasks.
+In each iteration of this loop, each provider is responsible for determining how many instances to start, and starting them.
+
+### Provisioning Workers
+
+The provisioning process begins by calling each provider's `initiate` method.
+This is similar to `setup`, but is not called in the web service.
+
+Similarly, when the provisioning process stops cleanly, it calls each provider's `terminate` method.
+Of course, clean stops are not the norm -- typically the process runs continuously until it crashes or is terminated -- so providers should not rely on `terminate` to clean up resources in production.
```

Comment:
I guess this isn't technically true since services are restarted at least once a day but we don't handle that by cleaning up.

Reply:
Yeah, even in that case I think the process is just SIGTERM'd, right?  So `terminate` wouldn't be called?

Reply:
Yeah, I believe that's what happens. We could start catching SIGTERM and doing cleanup on it, but we've never needed to.